crates/uv-resolver/src/lock.rs

```diff
@@ -653,128 +702,257 @@ impl Distribution {
+                    self.to_source_dist(workspace_root)?
+                {
+                    // When resolving from a lockfile all sources are equally compatible.
+                    let compat = SourceDistCompatibility::Compatible(HashComparison::Matched);
+                    let hash = self.sdist.as_ref().unwrap().hash().map(|h| h.0.clone());
+                    prioritized_dist.insert_source(sdist, hash, compat);
+                };
+
+                // Add any wheels.
+                for wheel in &self.wheels {
+                    let hash = wheel.hash.as_ref().map(|h| h.0.clone());
+                    let wheel = wheel.to_registry_dist(url, &self.id.source);
+                    let compat =
+                        WheelCompatibility::Compatible(HashComparison::Matched, None, None);
+                    prioritized_dist.insert_built(wheel, hash, compat);
                 }
-                Source::Git(url, git) => {
-                    // Reconstruct the `GitUrl` from the `GitSource`.
-                    let git_url =
-                        uv_git::GitUrl::new(url.clone(), GitReference::from(git.kind.clone()))
-                            .with_precise(git.precise);
-
-                    // Reconstruct the PEP 508-compatible URL from the `GitSource`.
-                    let url = Url::from(ParsedGitUrl {
-                        url: git_url.clone(),
-                        subdirectory: git.subdirectory.as_ref().map(PathBuf::from),
+
+                prioritized_dist
+            }
+            _ => return Ok(None),
+        };
+
+        Ok(Some(prioritized_dist))
+    }
+
+    /// Convert the [`Distribution`] to [`Metadata`] that can be used for resolution.
+    fn into_metadata(self, workspace_root: &Path) -> Result<Metadata, LockError> {
+        let name = self.name().clone();
+        let version = self.id.version.clone();
+        let provides_extras = self.optional_dependencies.keys().cloned().collect();
+
+        let mut dependency_extras = HashMap::new();
+        let mut requires_dist = self
+            .dependencies
+            .into_iter()
+            .filter_map(|dep| {
+                dep.into_requirement(workspace_root, &mut dependency_extras)
+                    .transpose()
+            })
+            .collect::<Result<Vec<_>, LockError>>()?;
+
+        // Denormalize optional dependencies.
+        for (extra, deps) in self.optional_dependencies {
+            for dep in deps {
+                if let Some(mut dep) =
+                    dep.into_requirement(workspace_root, &mut dependency_extras)?
+                {
+                    // Add back the extra marker expression.
+                    let marker = MarkerTree::Expression(MarkerExpression::Extra {
+                        operator: ExtraOperator::Equal,
+                        name: extra.clone(),
                     });
+                    match dep.marker {
+                        Some(ref mut tree) => tree.and(marker),
```

Comment:
A little concerned about reconstructing the `MarkerTree` like this without knowing the original order, but because we remove it when reconstructing the lockfile I think it's fine.

Reply:
Yeah I think the metadata we're constructing here is potentially incorrect... But perhaps not in a way that matters? It's hard to reason about. Like, we won't list all the extras here so `Provides-Extras` will be wrong -- we'll only include the extras that were activated when resolving...

Reply:
Yeah it makes some sense to me that we're doing this. I think it's adding back what we do here?

https://github.com/astral-sh/uv/blob/b677a06abad58868e226e19e722842f9bd8ec816/crates/pypi-types/src/metadata.rs#L254

Which reminds me: is it possible to simplify this code a bit via `MarkerTree::with_extra_marker`?

https://github.com/astral-sh/uv/blob/b677a06abad58868e226e19e722842f9bd8ec816/crates/pep508-rs/src/lib.rs#L412-L435

Reply:
Looks like that's `Requirement::with_extra_marker` not `MarkerTree::with_extra_marker`, but I can move it.

Reply:
Looks like that's `pep508_rs::Requirement::with_extra_marker` and we need `pypi_types::Requirement::with_extra_marker`. Ideally this could be on `MarkerTree` but the `Option` gets in the way..

Reply:
I mentioned this on Discord, but there's at least one problem with the partial metadata here. If we have some package in the lockfile (like `flask`), but it wasn't requested with an extra initially, we won't write that non-requested extra to the lockfile. If someone then comes in and asks for `flask[dotenv]`, we won't have the list of dependencies that come with `flask[dotenv]` -- we just don't store that in the lockfile at all.

So we need some solution for this. Note that non-existent extras are _not_ an error in the resolver; we just warn, mostly because extras can vary over time.